src/typeSystem/typeVariables.ts

```diff
@@ -0,0 +1,489 @@
+import { is, last, List, match, when } from '../extensions'
+import { Assignment, Body, Class, Closure, Describe, Environment, Expression, Field, If, Import, Literal, Method, Module, NamedArgument, New, Node, Package, Parameter, Program, Reference, Return, Self, Send, Singleton, Super, Test, Throw, Try, Variable } from '../model'
+import { ANY, AtomicType, ELEMENT, RETURN, TypeSystemProblem, VOID, WollokAtomicType, WollokClosureType, WollokMethodType, WollokModuleType, WollokParameterType, WollokParametricType, WollokType, WollokUnionType } from './wollokTypes'
+
+const { assign } = Object
+
+const tVars = new Map<Node, TypeVariable>()
+
+export function newTypeVariables(env: Environment): Map<Node, TypeVariable> {
+  tVars.clear()
+  createTypeVariables(env)
+  return tVars
+}
+
+export function newSynteticTVar(node?: Node): TypeVariable {
+  return doNewTVarFor(node?.copy() ?? Closure({ code: 'Param type' })) // Using new closure as syntetic node. Is good enough? No.
+    .beSyntetic()
+}
+
+export function typeVariableFor(node: Node): TypeVariable {
+  const tVar = tVars.get(node)
+  if (!tVar) return newTVarFor(node)
+  return tVar
+}
+
+
+function newTVarFor(node: Node) {
+  const newTVar = doNewTVarFor(node)
+  let annotatedVar = newTVar // By default, annotations reference the same tVar
+  if (node.is(Method)) {
+    const parameters = node.parameters.map(p => createTypeVariables(p)!)
+    annotatedVar = newSynteticTVar(node) // But for methods, annotations reference to return tVar
+    newTVar.setType(new WollokMethodType(annotatedVar, parameters, annotatedVariableMap(node)), false)
+  }
+  if (node.is(Singleton) && node.isClosure()) {
+    const methodApply = node.methods.find(_ => _.name === '<apply>')!
+    const parameters = methodApply.parameters.map(p => typeVariableFor(p))
```

Comment:
```suggestion
    const parameters = methodApply.parameters.map(typeVariableFor)
```

de nuevo, no estoy 100% seguro pero así podríamos evitar tener un parámetro de una sola letra y la burocracia de llamarlo.

Reply:
Esto lo banco. Nombras los parámetros me da un toque de paja cuando es evidente lo que es: por ejemplo un `map` en una colección que se llama `parameters` jajja

Igual te banco Dodain!